### ISSUE TITLE
When using "AND Logic" in "requires", the job may not be started in rare cases

### ISSUE THREAD

**Issue by wahapo** [screwdriver-cd/screwdriver#2320]
**What happened**:
Build may not start even if AND condition set in requires is met.
<img width="300" src="https://user-images.githubusercontent.com/24538326/105322640-7c515500-5c0c-11eb-80d9-1c82637a50d7.png">

**What you expected to happen**:
It's a bug and I hope it's fixed.

**How to reproduce it**:
1. Create `screwdriver.yaml`.
    ```
    shared:
       image: node:12
      steps:
        - ls
    
    jobs:
      start:
        requires: [ ~pr, ~commit ]
    
      sub-1:
        requires: [ ~start ]
    
      sub-2:
        requires: [ ~start ]
    
      sub-3:
        requires: [ ~start ]
    
      end:
        requires: [ sub-1, sub-2, sub-3 ]
    ```
2. Start `start` job.
3. Two or more of `sub-1`, `sub-2`, and `sub-3` complete and call `PUT /v4/build/{id}` API at about the same time. _(The probability that multiple jobs will be completed almost simultaneously is quite low.)_ ...*
4. Even if all jobs of `sub-1`, `sub-2`, and `sub-3` are completed, the `end` job will not start.

**_Note: * How to reproduce `3`?_**
The following method can sometimes reproduce it.
1. Remove these lines.
https://github.com/screwdriver-cd/screwdriver/blob/82100989eed8bee355df8bf4af23732d55d2337e/plugins/builds/update.js#L94-L100
2. By executing the following code, you can mimic the behavior of three jobs calling the `PUT /v4/build/{id}` API at the same time.
    ```
    JOBID1=<JobId of sub-1>
    JOBID2=<JobId of sub-2>
    JOBID3=<JobId of sub-3>
    TOKEN="<token string>"
    
    curl -X PUT "http://localhost:9001/v4/builds/${JOBID1}" -H  "accept: application/json" -H  "Authorization: ${TOKEN}" -H  "Content-Type: application/json" -d "{  \"meta\": {},  \"status\": \"SUCCESS\",  \"statusMessage\": \"Build failed due to infrastructure error\",  \"stats\": {    \"queueEnterTime\": \"2021-01-21T11:39:50\",    \"blockedStartTime\": \"2021-01-21T11:39:50\",    \"imagePullStartTime\": \"2021-01-21T11:39:50\",    \"hostname\": \"string\"  }}" & \
    curl -X PUT "http://localhost:9001/v4/builds/${JOBID2}" -H  "accept: application/json" -H  "Authorization: ${TOKEN}" -H  "Content-Type: application/json" -d "{  \"meta\": {},  \"status\": \"SUCCESS\",  \"statusMessage\": \"Build failed due to infrastructure error\",  \"stats\": {    \"queueEnterTime\": \"2021-01-21T11:39:50\",    \"blockedStartTime\": \"2021-01-21T11:39:50\",    \"imagePullStartTime\": \"2021-01-21T11:39:50\",    \"hostname\": \"string\"  }}" & \
    curl -X PUT "http://localhost:9001/v4/builds/${JOBID3}" -H  "accept: application/json" -H  "Authorization: ${TOKEN}" -H  "Content-Type: application/json" -d "{  \"meta\": {},  \"status\": \"SUCCESS\",  \"statusMessage\": \"Build failed due to infrastructure error\",  \"stats\": {    \"queueEnterTime\": \"2021-01-21T11:39:50\",    \"blockedStartTime\": \"2021-01-21T11:39:50\",    \"imagePullStartTime\": \"2021-01-21T11:39:50\",    \"hostname\": \"string\"  }}"
    ```

**Cause Investigation**
If `PUT /v4/build/{id}` is called multiple times at the same time and [`createInternalBuild`](https://github.com/screwdriver-cd/screwdriver/blob/82100989eed8bee355df8bf4af23732d55d2337e/plugins/builds/index.js#L918) is executed at the same time, multiple rows will be created in the DB. In the following case, three `end` jobs are created, and none of them will be started.
```
| parentBuildId | parentBuilds                                                                  |
|---------------|-------------------------------------------------------------------------------|
| 82            | {"1":{"eventId":35,"jobs":{"sub-1":82,"sub-2":null,"sub-3":null,"start":81}}} |
| 83            | {"1":{"eventId":35,"jobs":{"sub-1":null,"sub-2":83,"sub-3":null,"start":81}}} |
| 84            | {"1":{"eventId":35,"jobs":{"sub-1":null,"sub-2":null,"sub-3":84,"start":81}}} |
```

In the following case, as a result of `sub-2` and `sub-3` calling `PUT /v4/build/{id}` API at the same time, the data of `sub-2` is lost from the `parentEventId` of the `end` job. This is probably because `sub-3` runs [`updateParentBuilds`](https://github.com/screwdriver-cd/screwdriver/blob/82100989eed8bee355df8bf4af23732d55d2337e/plugins/builds/index.js#L921-L927) slightly later than `sub-2`.
```
| parentBuildId | parentBuilds                                                                  |
|---------------|-------------------------------------------------------------------------------|
| [63, 61]      | {"1":{"eventId":31,"jobs":{"sub-1":61,"sub-2":null,"sub-3":63,"start":60}}}   |
```

**Comment by jithine** [comment]
@wahapo This is the same issue as this https://github.com/screwdriver-cd/screwdriver/issues/2071

There is indeed a race condition, and we need a holistic fix for this 😞

**Comment by redonkulus** [comment]
@wahapo @jithine I believe I found this same issue with the OR logic as well. Can you please test that scenario too?

**Comment by wahapo** [comment]
@redonkulus I think it is a different phenomenon in `OR` case.

I've tried.
https://cd.screwdriver.cd/pipelines/6441/events/606628

`sub-{1,2,3,4}` jobs create four `end` jobs successfully, but only one job starts. In addition, only one of the QUEUED jobs is shown in the pipeline graph.
- https://cd.screwdriver.cd/pipelines/6441/builds/674134 (SUCCESS, Not visible in pipeline graph)
- https://cd.screwdriver.cd/pipelines/6441/builds/674135 (QUEUED, Not visible in pipeline graph)
- https://cd.screwdriver.cd/pipelines/6441/builds/674136 (QUEUED, Not visible in pipeline graph)
- https://cd.screwdriver.cd/pipelines/6441/builds/674137 (QUEUED, Visible in pipeline graph)

**Comment by wahapo** [comment]
Close this issue because it's a duplicate of https://github.com/screwdriver-cd/screwdriver/issues/2071
